datasource/src/main/scala/quasar/physical/mongo/Mongo.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+import quasar.api.resource.{ResourceName, ResourcePath}
+import quasar.connector.{MonadResourceErr, ResourceError}
+import quasar.physical.mongo.MongoResource.{Database, Collection}
+
+import cats.effect.{ConcurrentEffect, IO}
+import cats.syntax.eq._
+import fs2.Stream
+import fs2.concurrent._
+import org.bson.{Document => _, _}
+import org.mongodb.scala._
+
+class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: MongoClient) {
+  import Mongo._
+
+  val F: ConcurrentEffect[F] = ConcurrentEffect[F]
+
+  def databases: Stream[F, Database] =
+    observableAsStream(client.listDatabaseNames).map(Database(_))
+
+  def databaseExists(database: Database): Stream[F, Boolean] =
+    databases.exists(_ === database)
+
+  def collections(database: Database): Stream[F, Collection] = for {
+    dbExists <- databaseExists(database)
+    res <- if (!dbExists) { Stream.empty } else {
+      observableAsStream(client.getDatabase(database.name).listCollectionNames()).map(Collection(database, _))
+    }
+  } yield res
+
+  def collectionExists(collection: Collection): Stream[F, Boolean] =
+    collections(collection.database).exists(_ === collection)
+
+  def findAll(collection: Collection): Stream[F, BsonValue] = for {
+    colExists <- collectionExists(collection)
+    res <- if (colExists) {
+      observableAsStream(
+        client.getDatabase(collection.database.name).getCollection(collection.name).find[BsonValue]()
+      )
+    } else {
+      Stream.raiseError(
+        ResourceError.throwableP(
+          ResourceError.pathNotFound(
+            ResourcePath.root() / ResourceName(collection.database.name) / ResourceName(collection.name)
+          )
+        )
+      )
+    }
+  } yield res
+
+  def close: F[Unit] = F.delay(client.close())
+
+}
+object Mongo {
+  def observableAsStream[F[_], A](obs: Observable[A])(implicit F: ConcurrentEffect[F]): Stream[F, A] = {
```

Comment:
A couple issues here

1) Unbounded queues should be avoided so that we can control memory usage, we need to add backpressure via managing calls to `subscription.request(n)` in `onSubscribe` and `onNext`. I'd start with just requesting a single item at a time, we can experiment with batches if performance seems poor.

2) We need to handle early termination of the `Stream` via `subscription.unsubscribe()`. As-is, if a user did a `take(1)` on the returned stream, we'd still stream the entire result, filling up our unbounded queue.

Reply:
This is super-important and needs to block the merge.